### PR TITLE
[dagster-airlift] lazy loading

### PR DESCRIPTION
**Pull request by dpeng817**
Re-implement airflow defs builder on top of lazy defs loading abstractions.

Added a test for caching behavior.
## Changelog
`NOCHANGELOG`

### CODE REVIEW

**Comment by dpeng817** [comment]
* **#24540** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24540?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/conftest.py [hunk 1]
```diff
@@ -122,3 +129,28 @@ def assert_dependency_structure_in_assets(
...
+        self,
+        metadata: Mapping[str, Any] = {},
+        load_type: DefinitionsLoadType = DefinitionsLoadType.INITIALIZATION,
+    ):
+        super().__init__(load_type=load_type)
+        self._metadata = metadata
+
+    @property
+    def reconstruction_metadata(self) -> Mapping[str, Any]:
+        return self._metadata
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/conftest.py:145]: You shouldn't need this. You can construct a `DefinitionsLoadContext` directly and substitute the right values.

Comment (dpeng817) [reply examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/conftest.py:145]: DefinitionsLoadContext requires a repository load data be passed in, which is quite tedious to construct. I found this easier

Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/conftest.py]: `RepositoryLoadData({}, metadata)`

?

File: examples/experimental/dagster-airlift/dagster_airlift/core/defs_from_airflow.py [hunk 2]
```diff
@@ -1,60 +1,492 @@
...
-    airflow_instance: AirflowInstance,
-    cache_polling_interval: int = DEFAULT_POLL_INTERVAL,
-    defs: Optional[Definitions] = None,
-    # This parameter will go away once we can derive the migration state from airflow itself, using our built in utilities.
-    # Alternatively, we can keep it around to let people override the migration state if they want.
-    migration_state_override: Optional[AirflowMigrationState] = None,
+def load_airflow_definitions(
+    context: DefinitionsLoadContext,
+    instance: AirflowInstance,
+    defs: Definitions = Definitions(),
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/defs_from_airflow.py:54]: don't do this

Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift/core/defs_from_airflow.py:54]: https://docs.python-guide.org/writing/gotchas/

**Review by schrockn — CHANGES_REQUESTED** [review]
Per zoom convo, add adapter layer so that examples remaining unchanged.

**Comment by dpeng817** [comment]
* **#24569** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24569?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24556** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24556?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24540** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24540?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by dpeng817** [comment]
* **#24569** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24569?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24556** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24556?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24540** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24540?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by dpeng817** [comment]
* **#24569** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24569?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24556** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24556?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24540** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24540?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by dpeng817** [comment]
* **#24569** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24569?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24556** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24556?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24540** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24540?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py [hunk 3]
```diff
@@ -41,6 +47,123 @@
...
+    else:
+        context = DefinitionsLoadContext(load_type=DefinitionsLoadType.INITIALIZATION)
+    metadata_key = f"{AIRFLOW_SOURCE_METADATA_KEY_PREFIX}/{airflow_instance.name}"
+    definitions_metadata = _get_or_create_definitions_metadata(
+        context, metadata_key, airflow_instance, defs
+    )
+    return definitions_from_metadata(
+        context, definitions_metadata, defs, airflow_instance, sensor_minimum_interval_seconds
+    )
+
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py:70]: This is a bit messy, but we should be able to refactor it quickly over a lightweight, reusable pattern.

File: examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py [hunk 4]
```diff
@@ -41,6 +47,123 @@
...
+def definitions_from_metadata(
+    context: DefinitionsLoadContext,
+    definitions_metadata: "CacheableSpecs",
+    defs: Definitions,
+    airflow_instance: AirflowInstance,
+    sensor_minimum_interval_seconds: int,
+) -> Definitions:
+    new_assets_defs = [
+        external_asset_from_spec(
+            dag_spec.to_asset_spec(additional_metadata={}),
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py:126]: I'd like a convention where we suffix cacheable spes with `cacheable_spec`. A little more verbose but much more clear IMO.

File: examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py [hunk 5]
```diff
@@ -41,6 +47,123 @@
...
+        asset_checks=defs.asset_checks if defs else None,
+        sensors=[airflow_sensor, *defs.sensors] if defs and defs.sensors else [airflow_sensor],
+        schedules=defs.schedules if defs else None,
+        jobs=defs.jobs if defs else None,
+        executor=defs.executor if defs else None,
+        loggers=defs.loggers if defs else None,
+        resources=defs.resources if defs else None,
+    ).with_reconstruction_metadata(
+        {
+            f"{AIRFLOW_SOURCE_METADATA_KEY_PREFIX}/{airflow_instance.name}": serialize_value(
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py:160]: nit: would prefer this string construction wrapper in a common function

File: python_modules/dagster/dagster/_core/definitions/definitions_loader.py [hunk 6]
```diff
@@ -64,6 +64,11 @@ def get(cls) -> "DefinitionsLoadContext":
             )
         return DefinitionsLoadContext._instance
 
+    @classmethod
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/definitions_loader.py:67]: @smackesey  this API should be structured so this is not necessary. Recommendation is `get` defaults to an initializing context with no load data.

**Review by schrockn — DISMISSED** [review]
Looks good overall. I want alignment with @smackesey on case and pattern when the context has not explicitly been set.

File: python_modules/dagster/dagster/_core/definitions/definitions_loader.py (hunk 6 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/definitions_loader.py:67]: Yep finding a need for that myself PR up in a minute.

Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/definitions_loader.py:67]: https://github.com/dagster-io/dagster/pull/24626

File: examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py (hunk 3 above)
Comment (dpeng817) [reply examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py]: yea esp now that sean's changes landed, i think this feels a lot better.

File: examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py [hunk 7]
```diff
@@ -41,6 +47,119 @@
...
 
+@suppress_dagster_warnings
+def build_defs_from_airflow_instance(
+    *,
+    airflow_instance: AirflowInstance,
+    defs: Optional[Definitions] = None,
+    sensor_minimum_interval_seconds: int = DEFAULT_AIRFLOW_SENSOR_INTERVAL_SECONDS,
+) -> Definitions:
+    defs = defs or Definitions()
+    # Probably a temp workaround?
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py:58]: rm?

**Review by schrockn — APPROVED** [review]

**Pull request merged**